Implement fingerprinting countermeasures for querying the permission state of the Notifications API

RupinMittal opened the pull request:
#### a742d4b71e27e2f1e82e90c3b26150aa0366fdcd
<pre>
Implement fingerprinting countermeasures for querying the permission state of the Notifications API
<a href="https://bugs.webkit.org/show_bug.cgi?id=243215">https://bugs.webkit.org/show_bug.cgi?id=243215</a>

Reviewed by Chris Dumez and Sihui Liu.

To prevent user fingerprinting, a website should not be able to find out that a user has denied
permission to use the Notifications API unless the website actually wants to use that API. So if
a user has denied permission, and the website calls Permissions::query() to find out the permission
state of the Notifications API, it will receive &quot;prompt&quot; rather than &quot;denied&quot;. Permissions::query()
will return &quot;denied&quot; only after the website requests permission to use the API. If the permission
state is &quot;prompt&quot; or &quot;granted&quot;, the true result is always returned. Notification::permission()
will have this same behavior (in both WebKit2 and WebKitLegacy). Additionally, all calls to
Notification::requestPermission will now go to the UIProcess, and thus, to the client.

In order to test this change, modifications were made to TestRunner. Some existing layout tests were
also modified to expect this behavior.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/notifications/permission/notification-permission-query-expected.txt: Added.
* LayoutTests/http/tests/notifications/permission/notification-permission-query-fingerprinting-expected.txt: Added.
* LayoutTests/http/tests/notifications/permission/notification-permission-query-fingerprinting.html: Added.
* LayoutTests/http/tests/notifications/permission/notification-permission-query.html: Added.
* LayoutTests/http/tests/notifications/show-expected.txt:
* LayoutTests/http/tests/notifications/show.html:
* LayoutTests/http/tests/push-api/subscribe-deny-permissions-expected.txt:
* LayoutTests/http/tests/push-api/subscribe-deny-permissions.html:
* LayoutTests/http/tests/workers/service/shownotification-denied-expected.txt:
* LayoutTests/http/tests/workers/service/shownotification-denied.html:
* LayoutTests/platform/wk2/TestExpectations:
* Source/WebCore/Modules/notifications/NotificationController.h:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageClearNotificationPermissionState):
* Source/WebKit/UIProcess/API/C/WKPage.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::queryPermission):
(WebKit::WebPageProxy::clearNotificationPermissionState):
(WebKit::WebPageProxy::requestNotificationPermission):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/Notifications/NotificationPermissionRequestManager.cpp:
(WebKit::NotificationPermissionRequestManager::startRequest):
* Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
(WebKit::WebNotificationClient::requestPermission):
(WebKit::WebNotificationClient::checkPermission):
(WebKit::WebNotificationClient::clearNotificationPermissionState):
* Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::clearNotificationPermissionState):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKitLegacy/mac/WebCoreSupport/WebNotificationClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebNotificationClient.mm:
(WebNotificationClient::clearNotificationPermissionState):
(WebNotificationClient::requestPermission):
(WebNotificationClient::checkPermission):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _clearNotificationPermissionState]):
* Source/WebKitLegacy/mac/WebView/WebViewPrivate.h:
* Tools/DumpRenderTree/TestRunner.h:
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(runTest):
* Tools/DumpRenderTree/mac/TestRunnerMac.mm:
(TestRunner::clearNotificationPermissionState):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NotificationAPI.mm:
(TestWebKitAPI::runRequestPermissionTest):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::queryPermission):
(WTR::TestController::handleQueryPermission):
(WTR::TestController::resetStateToConsistentValues):
(WTR::TestController::decidePolicyForNotificationPermissionRequest):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/WebNotificationProvider.cpp:
(WTR::WebNotificationProvider::permissionState):
* Tools/WebKitTestRunner/WebNotificationProvider.h:

Canonical link: <a href="https://commits.webkit.org/252969@main">https://commits.webkit.org/252969@main</a>
</pre>

cdumez commented on LayoutTests/http/tests/notifications/notification-permission-query.html:
```diff
@@ -0,0 +1,73 @@
+
+function defaultTest()
+{
+    testRunner.removeAllWebNotificationPermissions();
```
What's this? This shouldn't be needed. We don't rely on tests to clear state at the beginning or at the end of the test. It is the WebKitTestRunner's job.

RupinMittal replied:
Fixed!

cdumez commented on LayoutTests/http/tests/notifications/notification-permission-query.html:
```diff
@@ -0,0 +1,73 @@
+function grantTest() 
+{    
+    internals.withUserGesture(() => {
+        testRunner.grantWebNotificationPermission(testURL);
```
Why is this needed? I don't think we should need this?
By default, when you request permission, it gets granted.

RupinMittal replied:
Fixed!

cdumez commented on LayoutTests/http/tests/notifications/notification-permission-query.html:
```diff
@@ -0,0 +1,73 @@
+function denyTest() 
+{
+    internals.withUserGesture(() => {
+        testRunner.denyWebNotificationPermission(testURL);
```
I think we should call the OnPrompt function here. Looks like we're testing the user granting / denying from the prompt.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8856,8 +8856,12 @@ void WebPageProxy::queryPermission(const ClientOrigin& clientOrigin, const Permi
+
+        // Ensure that the true permission state of the Notifications API is returned if
+        // this topOrigin has requested permission to use the Notifications API previously.
+        WebCore::SecurityOriginData topOrigin = clientOrigin.topOrigin;
```
No need for this local variable...

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8856,8 +8856,12 @@ void WebPageProxy::queryPermission(const ClientOrigin& clientOrigin, const Permi
+        // Ensure that the true permission state of the Notifications API is returned if
+        // this topOrigin has requested permission to use the Notifications API previously.
+        WebCore::SecurityOriginData topOrigin = clientOrigin.topOrigin;
+        if (m_notificationsRequesters.contains(topOrigin))
```
... just pass `clientOrigin.topOrigin` here.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -9071,6 +9075,11 @@ void WebPageProxy::showMediaControlsContextMenu(FloatRect&& targetFrame, Vector<
     auto origin = API::SecurityOrigin::createFromString(originString);
+
+    // Add origin to list of origins that have requested permission to use the Notifications API.
+    WebCore::SecurityOriginData securityOriginData = origin.get().securityOrigin();
```
No need for this local variable.

RupinMittal replied:
Fixed!

cdumez commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -360,8 +361,28 @@ void TestController::setIsMediaKeySystemPermissionGranted(bool granted)
+
+void TestController::handleQueryPermission(WKStringRef string, WKSecurityOriginRef securityOrigin, WKQueryPermissionResultCallbackRef callback)
+{
+    if (toWTFString(string) == toWTFString(std::string("notifications"))) {
```
Should just be:
```
if (toWTFString(string) == "notifications"_s) {
```

RupinMittal replied:
Fixed!

cdumez commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -360,8 +361,28 @@ void TestController::setIsMediaKeySystemPermissionGranted(bool granted)
+void TestController::handleQueryPermission(WKStringRef string, WKSecurityOriginRef securityOrigin, WKQueryPermissionResultCallbackRef callback)
+{
+    if (toWTFString(string) == toWTFString(std::string("notifications"))) {
+        WKRetainPtr<WKStringRef> securityOriginString = adoptWK(WKSecurityOriginCopyToString(securityOrigin));
```
auto

RupinMittal replied:
I can make this change--but I'm curious--why would we use auto if we can be specific about the type?

RupinMittal replied:
Got it--it is standard Webkit style policy to use auto where possible. Fixed!

cdumez commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -360,8 +361,28 @@ void TestController::setIsMediaKeySystemPermissionGranted(bool granted)
+        WKRetainPtr<WKDictionaryRef> dictionary = m_webNotificationProvider.notificationPermissions();
+
+        auto value = WKDictionaryGetItemForKey(dictionary.get(), securityOriginString.get());
+        if (value) {
```
Could reduce scoping like so:
```
if (auto value = WKDictionaryGetItemForKey(dictionary.get(), securityOriginString.get())) {
```

RupinMittal replied:
Fixed!

cdumez commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -360,8 +361,28 @@ void TestController::setIsMediaKeySystemPermissionGranted(bool granted)
+
+        auto value = WKDictionaryGetItemForKey(dictionary.get(), securityOriginString.get());
+        if (value) {
+            bool permission = WKBooleanGetValue(static_cast<WKBooleanRef>(value));
```
Let's rename "permission" to "wasGranted"

RupinMittal replied:
Fixed!

cdumez commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -360,8 +361,28 @@ void TestController::setIsMediaKeySystemPermissionGranted(bool granted)
+        auto value = WKDictionaryGetItemForKey(dictionary.get(), securityOriginString.get());
+        if (value) {
+            bool permission = WKBooleanGetValue(static_cast<WKBooleanRef>(value));
+            if (permission == 1)
```
`if (wasGranted)`

RupinMittal replied:
Fixed!

cdumez commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -2688,11 +2709,19 @@ void TestController::decidePolicyForNotificationPermissionRequest(WKPageRef page
     auto originName = originUserVisibleName(origin);
-    if (m_notificationOriginsToDenyOnPrompt.contains(originName)) {
+
+    WKRetainPtr<WKStringRef> securityOriginString = adoptWK(WKSecurityOriginCopyToString(origin));
```
auto

RupinMittal replied:
Fixed!

cdumez commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -2688,11 +2709,19 @@ void TestController::decidePolicyForNotificationPermissionRequest(WKPageRef page
+    WKRetainPtr<WKStringRef> securityOriginString = adoptWK(WKSecurityOriginCopyToString(origin));
+    WKRetainPtr<WKDictionaryRef> dictionary = m_webNotificationProvider.notificationPermissions();
+    auto value = WKDictionaryGetItemForKey(dictionary.get(), securityOriginString.get());
+    bool permissionDenied = value ? !(WKBooleanGetValue(static_cast<WKBooleanRef>(value))) : false;
```
`bool permissionDenied = value && !WKBooleanGetValue(static_cast<WKBooleanRef>(value));`

RupinMittal replied:
Fixed!

cdumez commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -2688,11 +2709,19 @@ void TestController::decidePolicyForNotificationPermissionRequest(WKPageRef page
+    bool permissionDenied = value ? !(WKBooleanGetValue(static_cast<WKBooleanRef>(value))) : false;
+
+    if (m_notificationOriginsToDenyOnPrompt.contains(originName) || permissionDenied) {
+        m_webNotificationProvider.setPermission(toWTFString(securityOriginString.get()), false);
```
This is unnecessary when `permissionDenied` is true. I think we should do 2 separate if checks.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -9071,6 +9075,11 @@ void WebPageProxy::showMediaControlsContextMenu(FloatRect&& targetFrame, Vector<
+
+    // Add origin to list of origins that have requested permission to use the Notifications API.
+    WebCore::SecurityOriginData securityOriginData = origin.get().securityOrigin();
+    m_notificationsRequesters.add(securityOriginData);
```
Shouldn't we do this in the completion handler? Also, can the completion handler be called with "Prompt" (e.g. if there was no user interaction). If so, it seems we should only add to the HashMap if the permission is different than prompt, no?

RupinMittal replied:
We decided not to do this in the permission handler since WebPageProxy::requestPermission() is not called if Notification::requestPermission() is called without a user gesture. So adding an origin to the HashSet as soon as they call requestPermission after a user gesture is correct behavior.

cdumez commented on LayoutTests/http/tests/notifications/notification-permission-query.html:
```diff
@@ -0,0 +1,73 @@
+    testRunner.removeAllWebNotificationPermissions();
+    debug("No permission state is set");
+    navigator.permissions.query({ name: "notifications" }).then((status)=>{
+        window.permissionsQueryResult = status.state;
```
`window.` is unnecessary.

RupinMittal replied:
Fixed!

cdumez commented on LayoutTests/http/tests/notifications/notification-permission-query-fingerprinting.html:
```diff
@@ -0,0 +1,55 @@
+
+function preRequestTest() 
+{
+    internals.withUserGesture(() => {
```
Why is the user gesture needed here? You're not requesting permission.

RupinMittal replied:
Fixed!

szewai commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -360,8 +361,28 @@ void TestController::setIsMediaKeySystemPermissionGranted(bool granted)
+    TestController::singleton().handleQueryPermission(string, securityOrigin, callback);
+}
+
+void TestController::handleQueryPermission(WKStringRef string, WKSecurityOriginRef securityOrigin, WKQueryPermissionResultCallbackRef callback)
```
Why adding `handleQueryPermission` function? Is it used anywhere else?

RupinMittal replied:
handQueryPermission was added because TestController::queryPermission needs to access a private data member of TestController, but it is a static method, so it can't access it. One fix was to add a function like handleQueryPermission and the other was to add a getter function to return the data member. Adding handleQueryPermission seems to be the preferred method.

RupinMittal replied:
For updating Notification::permission() to have the same behavior as Permissions::query() to reduce fingerprinting--you're right. The Permissions API feature is enabled, so we need to ensure that these two functions continue to have the same behavior. I've updated Notification::permission() in a similar way to Permissions::query().

szewai reviewed (commented):
I think you need to update Notification.permission to not return true value until requestPermission is called too; otherwise the fingerprint issue still exists (i.e. website can know the denied state) and  Notification.permission will return different values as permissions.query().

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -9068,9 +9071,19 @@ void WebPageProxy::showMediaControlsContextMenu(FloatRect&& targetFrame, Vector<
     auto origin = API::SecurityOrigin::createFromString(originString);
+
+    // Add origin to list of origins that have requested permission to use the Notifications API.
+    m_notificationsRequesters.add(origin.get().securityOrigin());
```
`origin->securityOrigin()`

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
```diff
@@ -119,9 +123,18 @@ NotificationClient::Permission WebNotificationClient::checkPermission(ScriptExec
         resultPermission = WebProcess::singleton().supplement<WebNotificationManager>()->policyForOrigin(origin);
     });
+
+    if (!m_notificationsRequesters.contains(context->securityOrigin()->data()) && resultPermission == NotificationClient::Permission::Denied)
```
Maybe a comment here to explain why we're doing this. Like you did for the other code path.

RupinMittal replied:
Fixed!

cdumez commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -360,8 +361,27 @@ void TestController::setIsMediaKeySystemPermissionGranted(bool granted)
+{
+    if (toWTFString(string) == "notifications"_s) {
+        auto securityOriginString = adoptWK(WKSecurityOriginCopyToString(securityOrigin));
+        auto dictionary = adoptWK(m_webNotificationProvider.notificationPermissions());
```
This is a pretty bad bug. You shouldn't be adopting the returned dictionary since it didn't come from a copy/create function. You need to ref the dictionary if you want to extend its lifetime.

RupinMittal replied:
We looked at the code and decided not to change this since WebNotificationProvide::notificationPermissions() calls retain on m_permissions.get(), so the adopt is needed. But I've changed the implementation to not call this WebNotificationProvide::notificationPermissions().

cdumez commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -360,8 +361,27 @@ void TestController::setIsMediaKeySystemPermissionGranted(bool granted)
+
+        if (auto value = WKDictionaryGetItemForKey(dictionary.get(), securityOriginString.get())) {
+            bool permissionWasGranted = WKBooleanGetValue(static_cast<WKBooleanRef>(value));
+            if (permissionWasGranted == 1)
```
Like I said in my earlier review, no `== 1`

RupinMittal replied:
Fixed!

cdumez commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -2688,11 +2709,24 @@ void TestController::decidePolicyForNotificationPermissionRequest(WKPageRef page
     auto originName = originUserVisibleName(origin);
+
+    auto securityOriginString = adoptWK(WKSecurityOriginCopyToString(origin));
+    auto dictionary = adoptWK(m_webNotificationProvider.notificationPermissions());
```
adoptWK() is wrong here.

RupinMittal replied:
I changed the implementation to not call m_webNotificationProvider.notificationPermissions().

cdumez commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -2688,11 +2709,24 @@ void TestController::decidePolicyForNotificationPermissionRequest(WKPageRef page
+    auto securityOriginString = adoptWK(WKSecurityOriginCopyToString(origin));
+    auto dictionary = adoptWK(m_webNotificationProvider.notificationPermissions());
+    auto value = WKDictionaryGetItemForKey(dictionary.get(), securityOriginString.get());
+    bool permissionDenied = value && !(WKBooleanGetValue(static_cast<WKBooleanRef>(value)));
```
Too many parentheses here. Should be:
`bool permissionDenied = value && !WKBooleanGetValue(static_cast<WKBooleanRef>(value));`

RupinMittal replied:
Fixed (with a slight change since we no longer call `m_webNotificationProvider.notificationPermissions()`

szewai commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -2688,11 +2709,24 @@ void TestController::decidePolicyForNotificationPermissionRequest(WKPageRef page
+        return;
+    }
+
+    if (permissionDenied) {
```
This check can be moved to above `if (m_notificationOriginsToDenyOnPrompt.contains(originName)) {` so we don't need to check m_notificationOriginsToDenyOnPrompt if we already know it's denied

RupinMittal replied:
Fixed!

szewai commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -360,8 +361,27 @@ void TestController::setIsMediaKeySystemPermissionGranted(bool granted)
+        auto securityOriginString = adoptWK(WKSecurityOriginCopyToString(securityOrigin));
+        auto dictionary = adoptWK(m_webNotificationProvider.notificationPermissions());
+
+        if (auto value = WKDictionaryGetItemForKey(dictionary.get(), securityOriginString.get())) {
```
You call notificationPermissions to get a permission state multiple times; you may reduce the duplicate code by adding getPermission() function to WebNotificationProvider. 
Also if notificationPermissions() is not used anywhere, we may just drop it so that all public access is done via get/set functions

RupinMittal replied:
WebNotificationProvider::notificationPermissions() is called internally, so I can't drop it. But I reduced the duplicate code by making a separate function to get the permission state (if there is one).

szewai commented on Source/WebKit/UIProcess/API/C/WKPage.h:
```diff
@@ -276,6 +276,8 @@ WK_EXPORT void WKPagePostMessageToInjectedBundle(WKPageRef page, WKStringRef mes
 
 WK_EXPORT void WKPageSelectContextMenuItem(WKPageRef page, WKContextMenuItemRef item);
 
+WK_EXPORT void WKPageClearNotificationState(WKPageRef page);
```
I think it's more clear to name it `WKPageClearNotificationPermissionState`

RupinMittal replied:
Fixed!

szewai commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2134,6 +2134,8 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>
     void updateIconForDirectory(NSFileWrapper *, const String&);
 #endif
 
+    void clearNotificationState();
```
Same here; it's more clear if it's named `clearNotificationPermissionState`

RupinMittal replied:
Fixed!

szewai commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2819,6 +2821,8 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>
 
     HashSet<WebEditCommandProxy*> m_editCommandSet;
 
+    HashSet<WebCore::SecurityOriginData> m_notificationsRequesters;
```
Let's be more specific by naming it `m_notificationPermissionRequesters`

RupinMittal replied:
Fixed!

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.h:
```diff
@@ -43,6 +44,7 @@ class WebNotificationClient final : public WebCore::NotificationClient {
 public:
     WebNotificationClient(WebPage*);
     virtual ~WebNotificationClient();
+    void clearNotificationState();
```
`clearNotificationPermissionState`

RupinMittal replied:
Fixed!

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.h:
```diff
@@ -51,7 +53,8 @@ class WebNotificationClient final : public WebCore::NotificationClient {
     WebCore::NotificationClient::Permission checkPermission(WebCore::ScriptExecutionContext*) final;
-    
+
+    HashSet<WebCore::SecurityOriginData> m_notificationsRequesters;
```
`m_notificationPermissionRequesters` or just `m_permissionRequesters`, because this is in the `WebNotificationClient` class which is all about notification

RupinMittal replied:
Fixed!

szewai commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -1557,6 +1558,8 @@ class WebPage : public API::ObjectImpl<API::Object::Type::BundlePage>, public IP
 
     void updateThrottleState();
 
+    void clearNotificationState();
```
`clearNotificationPermissionState`

RupinMittal replied:
Fixed!

cdumez commented on LayoutTests/TestExpectations:
```diff
@@ -103,6 +103,7 @@ fast/forms/enterkeyhint-attribute-values.html [ Skip ]
 fast/scrolling/keyboard-scrolling-distance-downArrow.html [ Skip ]
 fast/scrolling/keyboard-scrolling-distance-pageDown.html [ Skip ]
 fast/scrolling/unfocusing-page-while-keyboard-scrolling.html [ Skip ]
+http/tests/notifications/permission/ [ Skip ]
```
Please move so that it is right next to this line in this file:
`imported/w3c/web-platform-tests/permissions/ [ Skip ]`

RupinMittal replied:
Fixed!

cdumez commented on LayoutTests/platform/wk2/TestExpectations:
```diff
@@ -706,6 +706,7 @@ fast/images/gif-loop-count.html [ Pass ]
 
 # Only relevant for WebKit2
 http/tests/storage/storage-map-leaking.html [ Pass ]
+http/tests/notifications/permission/ [ Pass ]
```
Please move this right next to the following line in this file:
`imported/w3c/web-platform-tests/permissions/ [ Pass ]`

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/API/C/WKPage.cpp:
```diff
@@ -3188,3 +3188,10 @@ void WKPageDispatchActivityStateUpdateForTesting(WKPageRef pageRef)
     toImpl(pageRef)->dispatchActivityStateUpdateForTesting();
 }
+
+#if ENABLE(NOTIFICATIONS)
```
This isn't quite right, the `#if` check should be inside the function here since this is SPI and the function is not guarded in the header:
```
void WKPageClearNotificationPermissionState(WKPageRef pageRef)
 {
#if ENABLE(NOTIFICATIONS)
     toImpl(pageRef)->clearNotificationPermissionState();
#endif
 }
```

RupinMittal replied:
That makes sense. Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
```diff
@@ -119,9 +123,20 @@ NotificationClient::Permission WebNotificationClient::checkPermission(ScriptExec
+
+    // To reduce fingerprinting, if the origin has not requested permission to use the
+    // Notifications API, and the permission state is "denied", return "default" instead.
+    if (!m_notificationPermissionRequesters.contains(context->securityOrigin()->data()) && resultPermission == NotificationClient::Permission::Denied)
```
We should do the `resultPermission == NotificationClient::Permission::Denied` check first since it is cheaper:
```
if (resultPermission == NotificationClient::Permission::Denied && !m_notificationPermissionRequesters.contains(context->securityOrigin()->data()))
```

RupinMittal replied:
Fixed!

cdumez commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -1085,6 +1102,9 @@ bool TestController::resetStateToConsistentValues(const TestOptions& options, Re
     // Reset notification permissions
     m_webNotificationProvider.reset();
     m_notificationOriginsToDenyOnPrompt.clear();
+#if ENABLE(NOTIFICATIONS)
```
No need for this since this SPI is not guarded.

RupinMittal replied:
Fixed!

cdumez commented on Tools/WebKitTestRunner/WebNotificationProvider.cpp:
```diff
@@ -163,6 +163,22 @@ WKDictionaryRef WebNotificationProvider::notificationPermissions()
 
+// If the state is not stored in m_permissions, return false. Otherwise, return true and set
+// permissionIsGranted based on if the state is granted or denied.
+bool WebNotificationProvider::getPermissionState(WKSecurityOriginRef securityOrigin, bool& permissionIsGranted)
```
out-parameters are so old fashioned. In new code, I'd prefer if we returned an `std::optional<bool>` instead:
```
std::optional<bool> WebNotificationProvider::permissionState(WKSecurityOriginRef securityOrigin)
```

Also, per coding style, WebKit getters should not start with the prefix "get".

RupinMittal replied:
Fixed!

cdumez commented on Tools/WebKitTestRunner/WebNotificationProvider.cpp:
```diff
@@ -163,6 +163,22 @@ WKDictionaryRef WebNotificationProvider::notificationPermissions()
+{
+    auto securityOriginString = adoptWK(WKSecurityOriginCopyToString(securityOrigin));
+    if (auto value = WKDictionaryGetItemForKey(m_permissions.get(), securityOriginString.get())) {
+        if (WKBooleanGetValue(static_cast<WKBooleanRef>(value)))
```
`return WKBooleanGetValue(static_cast<WKBooleanRef>(value));`

RupinMittal replied:
Fixed!

cdumez commented on Tools/WebKitTestRunner/WebNotificationProvider.cpp:
```diff
@@ -163,6 +163,22 @@ WKDictionaryRef WebNotificationProvider::notificationPermissions()
+        return true;
+    }
+
+    return false;
```
`return std::nullopt;`

RupinMittal replied:
Fixed!

cdumez reviewed (commented):
Almost there.

cdumez reviewed (approved):
LGTM

bnham commented on Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
```diff
@@ -119,9 +123,20 @@ NotificationClient::Permission WebNotificationClient::checkPermission(ScriptExec
     callOnMainRunLoopAndWait([&resultPermission, origin = origin->data().toString().isolatedCopy()] {
         resultPermission = WebProcess::singleton().supplement<WebNotificationManager>()->policyForOrigin(origin);
     });
+
+    // To reduce fingerprinting, if the origin has not requested permission to use the
+    // Notifications API, and the permission state is "denied", return "default" instead.
+    if (resultPermission == NotificationClient::Permission::Denied && !m_notificationPermissionRequesters.contains(context->securityOrigin()->data()))
+        return NotificationClient::Permission::Default;
+
```
I would leave `WebNotificationClient::checkPermission` alone because within WebCore and WebKit, we need to know the true permission state and not the downgraded version of the permission state. For instance, `PushManager::subscribe` will check the permission state, and if the permission state is `Default`, it will show a permission dialog. With this change as-is, `Denied` will get mapped to `Default` in every new WebProcess for at least the first invocation of `checkPermission`. This means that users will see a permission prompt for origins where they've actively denied permissions within the browser settings pane. This is incorrect behavior.

You can add some other function that returns the downgraded permission state and have only `Notification::permission` and `PushManager::permissionState` call that new function.

cdumez replied:
I think we should test it but I personally don't think we have a bug. I think it is working as intended. I don't think the statement about double prompting is correct.

If the user previously denied, then WebCore thinks the permissions is "prompt" as is intended for privacy and will think it needs to requestPermission. It will request permission but it will NOT prompt since the UIProcess/client knows the policy is denied, it will simply reject the permission request.

This is the intended behavior of this code change for notification and the same should apply to push. Otherwise, we will end up with ways for fingerprint the notification permission state via the push API instead.

Again, we should check it but I am not convinced a code change will be needed.

szewai replied:
Yes, I think `requestPermission` will not prompt user when [user agent knows permission state](https://www.w3.org/TR/permissions/#dfn-prompt). The idea of sending requestPermission to UI process when web process already knows the state is denied (e.g. in `PushManager::subscribe`), is to let UI process know the page has requested permission and now it can reply true state to the other checkPermission requests. 
Also, it seems the `PermissionDescriptor` in [Push API](https://w3c.github.io/push-api/#permission) is different from [Notifications API](https://notifications.spec.whatwg.org/#permissions-integration), maybe we should not use the checkPermission directly and should let client decide whether they are the same?

bnham replied:
> If the user previously denied, then WebCore thinks the permissions is "prompt" as is intended for privacy and will think it needs to requestPermission. It will request permission but it will NOT prompt since the UIProcess/client knows the policy is denied, it will simply reject the permission request.

The current implementation of the request permission callback our browser doesn't follow this rule (see the implementation of the `_webView:requestNotificationPermissionForSecurityOrigin:decisionHandler:` delegate callback). It will unconditionally show a prompt if the current decision is not `UserNotificationPermissionPolicy::Allow`.

So I think if you want to commit this patch as-is, you need a follow-up patch in the browser to fix this behavior, or else users will get gratuitous permission prompts for Web Push when they've already actively denied that permission.

cdumez replied:
Even though we will call requestNotificationPermissionForSecurityOrigin again, Safari will NOT prompt again if it knows the answer (allow or deny). I still think we should verify manually before landing but I am fairly confident the patch will work as-is.

bnham replied:
Oops nevermind, I misread that part of the browser code. Looks like we call the decision handler in the browser properly when the permission is set to `UserNotificationPermissionPolicy::Deny`. So you're good to commit this.

RupinMittal replied:
Sounds good--I didn't make any changes besides fixing the API-test failure.

bnham commented on LayoutTests/http/tests/push-api/subscribe-deny-permissions.html:
```diff
@@ -11,10 +11,10 @@
+        await testDocumentPermissionState(registration, 'prompt');
         await testServiceWorkerSubscribe(registration, 'NotAllowedError');
-        await testDocumentSubscribeWithoutUserGesture(registration, 'NotAllowedError', 'User denied push permission');
+        await testDocumentSubscribeWithoutUserGesture(registration, 'NotAllowedError', 'Push notification prompting can only be done from a user gesture.');
```
This particular line in the test should probably not be changed after fixing the issue noted above.

cdumez replied:
The test does not have a user gesture, therefore it should not be able to distinguish denied from prompt. Otherwise it defeats the purpose of this privacy change. That said, we should probably add another check with a user gesture and make sure that it behaves correctly

cdumez replied:
Since the test doesn't have a user gesture, having this subscribe API throw a NotAllowedError with a message saying that the permission is denied is a clear violation of the privacy we're trying to enforce in this patch.

It looks like we're testing the "with user gesture" case below already. Therefore, I don't think we even need to add a test case.

Overall, I think my recommendation is to validate using Safari (manually) that the double prompting that Ben is concerned about does not happen.

RupinMittal replied:
Sounds good!--we checked and it seems to be fine. No changes were made to this.

szewai reviewed (changes requested):
The API test failure looks relevant; please fix it before landing.
(Looks like you may only need to update expectation since all requestPermission requests will go to UI process now.)

szewai reviewed (approved):
LGTM, let's wait for bots to finish before landing.

RupinMittal committed:
Implement fingerprinting countermeasures for querying the permission state of the Notifications API
https://bugs.webkit.org/show_bug.cgi?id=243215

Reviewed by Chris Dumez and Sihui Liu.

To prevent user fingerprinting, a website should not be able to find out that a user has denied
permission to use the Notifications API unless the website actually wants to use that API. So if
a user has denied permission, and the website calls Permissions::query() to find out the permission
state of the Notifications API, it will receive "prompt" rather than "denied". Permissions::query()
will return "denied" only after the website requests permission to use the API. If the permission
state is "prompt" or "granted", the true result is always returned. Notification::permission()
will have this same behavior (in both WebKit2 and WebKitLegacy). Additionally, all calls to
Notification::requestPermission will now go to the UIProcess, and thus, to the client.

In order to test this change, modifications were made to TestRunner. Some existing layout tests were
also modified to expect this behavior.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/notifications/permission/notification-permission-query-expected.txt: Added.
* LayoutTests/http/tests/notifications/permission/notification-permission-query-fingerprinting-expected.txt: Added.
* LayoutTests/http/tests/notifications/permission/notification-permission-query-fingerprinting.html: Added.
* LayoutTests/http/tests/notifications/permission/notification-permission-query.html: Added.
* LayoutTests/http/tests/notifications/show-expected.txt:
* LayoutTests/http/tests/notifications/show.html:
* LayoutTests/http/tests/push-api/subscribe-deny-permissions-expected.txt:
* LayoutTests/http/tests/push-api/subscribe-deny-permissions.html:
* LayoutTests/http/tests/workers/service/shownotification-denied-expected.txt:
* LayoutTests/http/tests/workers/service/shownotification-denied.html:
* LayoutTests/platform/wk2/TestExpectations:
* Source/WebCore/Modules/notifications/NotificationController.h:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageClearNotificationPermissionState):
* Source/WebKit/UIProcess/API/C/WKPage.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::queryPermission):
(WebKit::WebPageProxy::clearNotificationPermissionState):
(WebKit::WebPageProxy::requestNotificationPermission):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/Notifications/NotificationPermissionRequestManager.cpp:
(WebKit::NotificationPermissionRequestManager::startRequest):
* Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
(WebKit::WebNotificationClient::requestPermission):
(WebKit::WebNotificationClient::checkPermission):
(WebKit::WebNotificationClient::clearNotificationPermissionState):
* Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::clearNotificationPermissionState):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKitLegacy/mac/WebCoreSupport/WebNotificationClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebNotificationClient.mm:
(WebNotificationClient::clearNotificationPermissionState):
(WebNotificationClient::requestPermission):
(WebNotificationClient::checkPermission):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _clearNotificationPermissionState]):
* Source/WebKitLegacy/mac/WebView/WebViewPrivate.h:
* Tools/DumpRenderTree/TestRunner.h:
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(runTest):
* Tools/DumpRenderTree/mac/TestRunnerMac.mm:
(TestRunner::clearNotificationPermissionState):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NotificationAPI.mm:
(TestWebKitAPI::runRequestPermissionTest):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::queryPermission):
(WTR::TestController::handleQueryPermission):
(WTR::TestController::resetStateToConsistentValues):
(WTR::TestController::decidePolicyForNotificationPermissionRequest):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/WebNotificationProvider.cpp:
(WTR::WebNotificationProvider::permissionState):
* Tools/WebKitTestRunner/WebNotificationProvider.h:

Canonical link: https://commits.webkit.org/252969@main
